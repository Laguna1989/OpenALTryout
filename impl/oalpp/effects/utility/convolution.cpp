#include "convolution.hpp"
#include "dj_fft.h"
#include <algorithm>
#include <iostream>

namespace oalpp {
namespace effects {
namespace utility {

namespace {
std::vector<std::complex<float>> real2Complex(std::vector<float> const& realValues)
{
    std::vector<std::complex<float>> complexValues;
    std::transform(
        realValues.begin(), realValues.end(), std::back_inserter(complexValues), [](float v) {
            return std::complex<float> { v, 0.0f };
        });
    return complexValues;
}

std::vector<float> complex2real(std::vector<std::complex<float>> const& complexValues)
{
    float const gainFactor = 2.0f / sqrt(2.0f);
    std::vector<float> realValues;
    std::transform(complexValues.begin(), complexValues.end(), std::back_inserter(realValues),
        [gainFactor](std::complex<float> c) { return c.real() * gainFactor; });
    return realValues;
}

std::size_t getNextNSquare(std::size_t size)
{
    auto sizeN2 = 2;
    while (sizeN2 < size) {
        sizeN2 *= 2;
    }
    return sizeN2;
}

std::vector<std::complex<float>> doFFT(std::vector<float> samples)
{
    std::vector<std::complex<float>> samplesComplex = real2Complex(samples);
    samplesComplex.resize(getNextNSquare(samplesComplex.size()));
    return dj::fft1d(samplesComplex, dj::fft_dir::DIR_BWD);
}

std::vector<float> doFFT(std::vector<std::complex<float>> samples)
{
    -auto const transformed = dj::fft1d(samples, dj::fft_dir::DIR_FWD);
    return complex2real(transformed);
}

} // namespace

Convolution::Convolution(std::vector<float> const& kernel)
    : m_kernel { doFFT(kernel) }
{
}

std::vector<float> Convolution::process(std::vector<float> const& input)
{
    auto inputTransformed = doFFT(input);
    unsigned int biggerSize = std::max(inputTransformed.size(), m_kernel.size());

    std::vector<std::complex<float>> multipliedTransformed;
    multipliedTransformed.resize(biggerSize);

    for (auto i = 0U; i != biggerSize; ++i) {
        auto const kernelValue = i < m_kernel.size() ? m_kernel[i] : 0.0f;
        auto const inputValue = i < inputTransformed.size() ? inputTransformed[i] : 0.0f;
        multipliedTransformed[i] = kernelValue * inputValue;
    }

    return doFFT(multipliedTransformed);
}

} // namespace utility
} // namespace effects
} // namespace oalpp
