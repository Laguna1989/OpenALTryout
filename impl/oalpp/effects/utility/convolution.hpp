#ifndef OPENALPP_CONVOLUTION_HPP
#define OPENALPP_CONVOLUTION_HPP

#include "oalpp/effects/mono_effect_interface.hpp"
#include <complex>
#include <vector>

namespace oalpp {
namespace effects {
namespace utility {

class Convolution : public MonoEffectInterface {
public:
    explicit Convolution(std::vector<float> const& kernel);

    std::vector<float> process(std::vector<float> const& input) override;

private:
    std::vector<std::complex<float>> m_kernel {};
};

} // namespace utility
} // namespace effects
} // namespace oalpp

#endif // OPENALPP_CONVOLUTION_HPP
