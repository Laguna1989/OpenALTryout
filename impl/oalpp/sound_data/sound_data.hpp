#ifndef OPENALPP_SOUND_DATA_HPP
#define OPENALPP_SOUND_DATA_HPP

#include <oalpp/sound_data/sound_data_interface.hpp>

namespace oalpp {

class SoundData : public SoundDataInterface {
public:
    SoundData(std::vector<float> const& data, int sampleRate, int numberOfChannels);
    SoundData() = delete;
    ~SoundData() override = default;

    SoundData(SoundData const& data) = delete;
    SoundData(SoundData&& data) noexcept;

    SoundData& operator=(SoundData const&) = delete;
    SoundData& operator=(SoundData&&) noexcept;

    int getNumberOfChannels() const noexcept override;
    int getSampleRate() const noexcept override;
    std::vector<float> const& getSamples() const noexcept override;

private:
    std::vector<float> m_samples {};
    int m_sampleRate { 0 };
    int m_numberOfChannels { 0 };
};

} // namespace oalpp

#endif // OPENALPP_SOUND_DATA_HPP
