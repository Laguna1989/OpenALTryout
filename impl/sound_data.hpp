#ifndef OPENALPP_SOUND_DATA_HPP
#define OPENALPP_SOUND_DATA_HPP

#include "sound_data_interface.hpp"
#include <string>

namespace oalpp {

class SoundData : public SoundDataInterface {
public:
    explicit SoundData(std::string const& fileName);

    int getNumberOfChannels() const override;
    int getSampleRate() const override;
    std::vector<short> const& getSamples() const override;

private:
    std::vector<short> m_samples {};
    int m_sampleRate { 0 };
    int m_channels { 0 };
};

} // namespace oalpp

#endif // OPENALPP_SOUND_DATA_HPP
