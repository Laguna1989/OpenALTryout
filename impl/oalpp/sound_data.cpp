#include "sound_data.hpp"
#include "libnyquist/Decoders.h"
#include <utility>

namespace oalpp {

SoundData::SoundData(std::string const& fileName)
{
    auto fileData = std::make_unique<nqr::AudioData>();
    nqr::NyquistIO loader;
    loader.Load(fileData.get(), fileName);

    m_channels = fileData->channelCount;
    m_sampleRate = fileData->sampleRate;
    m_samples = std::move(fileData->samples);
}

int SoundData::getNumberOfChannels() const { return m_channels; }

int SoundData::getSampleRate() const { return m_sampleRate; }

std::vector<float> const& SoundData::getSamples() const { return m_samples; }

} // namespace oalpp
