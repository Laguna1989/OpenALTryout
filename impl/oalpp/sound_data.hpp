#ifndef OPENALPP_SOUND_DATA_INCLUDE_HPP
#define OPENALPP_SOUND_DATA_INCLUDE_HPP

#include "oalpp/sound_data/sound_data.hpp"
#include "oalpp/sound_data/sound_data_with_effect.hpp"

#endif // OPENALPP_SOUND_DATA_INCLUDE_HPP
