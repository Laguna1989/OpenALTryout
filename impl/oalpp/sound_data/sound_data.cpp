#include "sound_data.hpp"
#include <utility>

namespace oalpp {

SoundData::SoundData(std::vector<float> const& data, int sampleRate, int numberOfChannels)
    : m_samples { data }
    , m_sampleRate { sampleRate }
    , m_numberOfChannels { numberOfChannels }
{
}

SoundData::SoundData(SoundData&& data) noexcept
    : m_samples { std::move(data.m_samples) }
    , m_sampleRate { std::exchange(data.m_sampleRate, 0) }
    , m_numberOfChannels { std::exchange(data.m_numberOfChannels, 0) }
{
}

SoundData& SoundData::operator=(SoundData&& other) noexcept
{
    m_samples = std::move(other.m_samples);
    m_sampleRate = std::exchange(other.m_sampleRate, 0);
    m_numberOfChannels = std::exchange(other.m_numberOfChannels, 0);
    return *this;
}

int SoundData::getNumberOfChannels() const noexcept { return m_numberOfChannels; }

int SoundData::getSampleRate() const noexcept { return m_sampleRate; }

std::vector<float> const& SoundData::getSamples() const noexcept { return m_samples; }

} // namespace oalpp
