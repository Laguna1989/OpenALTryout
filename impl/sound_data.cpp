#include "sound_data.hpp"
#include "libnyquist/Decoders.h"

SoundData::SoundData(std::string const& fileName)
{
    auto fileData = std::make_shared<nqr::AudioData>();
    nqr::NyquistIO loader;
    loader.Load(fileData.get(), fileName);

    m_channels = fileData->channelCount;
    m_sampleRate = fileData->sampleRate;

    std::size_t size = fileData->samples.size();
    if (fileData->channelCount == 2) {
        if (fileData->samples.size() % 4 != 0) {
            // TODO figure out why there seems to be one sample missing when exporting with reaper.
            size += 2;
        }
    }

    m_samples.resize(size);

    std::transform(fileData->samples.begin(), fileData->samples.end(), m_samples.begin(),
        [](auto in) { return static_cast<short>(in * std::numeric_limits<short>::max()); });
}

int SoundData::getNumberOfChannels() const { return m_channels; }

int SoundData::getSampleRate() const { return m_sampleRate; }

std::vector<short> const& SoundData::getSamples() const { return m_samples; }
