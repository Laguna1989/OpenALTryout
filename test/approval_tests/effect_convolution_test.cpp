#include "ApprovalTests/ApprovalTests.hpp"
#include "catch2/catch.hpp"
#include "oalpp/effects/utility/convolution.hpp"
#include "oalpp/sound_data.hpp"
#include "approval_test_helpers.hpp"

TEST_CASE("convolution")
{
    std::string const fileName { "assets/test1.wav" };
    oalpp::SoundData buffer { fileName };
    oalpp::effects::utility::Convolution convolution { buffer.getSamples() };
    oalpp::SoundDataWithEffect soundWithEffect { buffer, convolution };

    // note: Due to platform dependent float behavior, the result needs to be converted to int for
    // the approval tests to work.
    ApprovalTests::Approvals::verifyAll(ApprovalTestHelpers::asInt(soundWithEffect.getSamples(), 100));
}
