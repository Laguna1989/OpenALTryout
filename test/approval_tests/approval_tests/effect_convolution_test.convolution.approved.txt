version https://git-lfs.github.com/spec/v1
oid sha256:21a1d1f53c69b83540b154b981885280b2399439eb25804f13f84d81b0f854e3
size 323799
